packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -3,59 +3,78 @@
- * @param MAX_TILES - Maximum number of tiles to calculate.
+ * Calculates the participants that should be renderd basis on the list of dominant
+ * speakers and currently renderd participants in a call.
+ * @param participants - Array containing all participants of a call.
+ * {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param dominantSpeakers - An array containing the userId of dominant speakers
+ * in a call in the order of their dominance. 0th index is the most dominant, 1st is the second most etc
+ * @param visibleParticipants - Array containing currently rendered (visible)
+ * participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param maxTiles - Maximum number of tiles to calculate.
  * @returns VideoGalleryRemoteParticipant[] {@link @azure/communication-react#VideoGalleryRemoteParticipant}
  */
 export const smartDominantSpeakerParticipants = (
   participants: VideoGalleryRemoteParticipant[] = [],
   dominantSpeakers: Array<string> = [],
   visibleParticipants: VideoGalleryRemoteParticipant[] = [],
-  MAX_TILES = 4
+  maxTiles = 4
 ): VideoGalleryRemoteParticipant[] => {
   if (!participants) return [];
 
-  // Only use the Max allowed dominant speakers.
-  const dominantSpeakerIds = dominantSpeakers.slice(0, MAX_TILES);
-
   // Don't apply any logic if total number of video streams is less than Max video streams.
-  if (participants.length <= MAX_TILES) return participants;
+  if (participants.length <= maxTiles) return participants;
 
-  // Initialize `visibleParticipants` if it is empty.
-  if (!visibleParticipants.length) {
-    visibleParticipants = participants.slice(0, MAX_TILES);
+  let currentParticipants = visibleParticipants.slice();
+  // Initialize `currentParticipants` if it is empty.
+  if (!currentParticipants.length) {
+    currentParticipants = participants.slice(0, maxTiles);
   }
 
-  const lastVisibleSpeakerIds = visibleParticipants.map((speaker) => speaker.userId);
+  // Only use the Max allowed dominant speakers.
+  const dominantSpeakerIds = Array.from(new Set(dominantSpeakers).values()).slice(0, maxTiles);
+  const lastVisibleSpeakerIds = currentParticipants.map((p) => p.userId);
   const newDominantSpeakerIds = dominantSpeakerIds.filter((id) => !lastVisibleSpeakerIds.includes(id));
 
   // Remove participants that are no longer dominant and replace them with new dominant speakers.
-  lastVisibleSpeakerIds.forEach((id, idx) => {
-    if (!dominantSpeakerIds.includes(id)) {
+  for (let index = 0; index < maxTiles; index++) {
+    const oldUserId = lastVisibleSpeakerIds[index];
+
+    if ((oldUserId && !dominantSpeakerIds.includes(oldUserId)) || oldUserId === undefined) {
       const replacement = newDominantSpeakerIds.shift();
-      if (replacement) {
-        lastVisibleSpeakerIds[idx] = replacement;
-      }
+      if (!replacement) break;
+      lastVisibleSpeakerIds[index] = replacement;
     }
+  }
```

Comment:
This is a complex algorithm, and small naming changes can make a big difference.

e.g., below this line, `lastVisibleSpeakerIds` no longer refers to last render. Also, it refers to participants, not speakers.

The length of `lastVisibleSpeakerIds` is bounded (by maxTiles), so I'd not be worried about creating a copy. Something like:

```
const newParticipantIDs = []
for (const i = 0; i < maxTiles; i++) {
    const oldUserId = lastVisibleSpeakerIds[index];
    if ((!oldUserId || !dominanSpeakerIDs.includes(oldUserId)) && newDominantSpeakerIds.length > 0 ) {
         newParticipantIDs.push(newDominantSpeakerIds.shift());
    } else if(oldUserId) {
        newParticipantIDs.push(oldUserId);
    } else {
        break;
    }
}
```

Reply:
I have tweaked the variable names.
But a lot of it is subjective.
I am finding the example you have pasted much harder to read TBH.